### ISSUE TITLE
custom linear solver for linprog

### ISSUE THREAD

**Issue by hakeemo** [scipy/scipy#11201]
**Is your feature request related to a problem? Please describe.**
In order to solve the sparse linear system in the interior-point method, external (non-scipy) solvers can be faster than the default ones ( eg. linear solvers from petsc4py). It would be nice to have an option to use any external solver.

**Describe the solution you'd like**
Add an argument to linprog to pass in a get_solver function which takes in the matrix M and returns a solve function.

**Comment by mdhaber** [comment]
Thanks for the suggestion @hakeemo. I've added this to the list of `linprog` enhancements in gh-9269. I think a good time to do this is when taking care of the `linprog` maintenance item from gh-9671:
> Simplify control over method='interior-point' solver: one option solver instead of separate boolean options cholesky, sym_pos, and lstsqr

I don't expect this to be terribly difficult. Would you be interested in helping to make this happen?

**Comment by hakeemo** [comment]
I would be happy to implement it @mdhaber. I am not sure what is the best way to do it. What I have in mind is:

1. add `get_solver` in the documentation as a possible entry in the options dictionary for linprog
2.  add  `get_solver=None`  as an argument in _linprog_ip,  _ip_hsd and _get_delta
3. if get_solver is not None, then just call get_solver instead of _get_solver.

Let me know if there are any issues with this and how you would like to simplify the controls. Should be straight forward.

**Comment by mdhaber** [comment]
Oops I missed this. Currently the interior point method chooses its solver based on several Boolean arguments to linprog (e.g. `sym_pos`). If one fails, it sets these Boolean values to effectively cycle to the next most robust (and next slower) solver. This should be refactored to cycle through a `list` of solver functions. Instead of Boolean arguments, an integer argument (not sure what the name should be) should index where in the list linprog should begin cycling from (for problems known to be numerically challenging). Also, linprog would accept an optional argument `solvers`, a list of solver functions that could be prepended to the internal list.
I can't look at the code right now to verify that this makes sense. Let me get back to you shortly!

**Comment by mdhaber** [comment]
@hakeemo btw your suggestion looks good and would be easier to implement, but we had planned to refactor how `_get_solver` works to make it more easily extendable anyway.

**Comment by mdhaber** [comment]
I was thinking that somewhere we would define a list of functions that, given a matrix `M`, return a solver, e.g.:
```
def sparse_super_lu(M):
   return sps.linalg.splu(M, permc_spec=permc_spec).solve
def sparse_least_squares(M):
   return lamba r: sps.linalg.lsqr(M, r)[0]  # or use def instead of lambda
solvers = [cholmod, sps.linalg.factorized, sparse_super_lu, sparse_least_squares]
```
The list above would be the default for sparse operation. Depending on the user's options (`sparse`, `lstsq`, `sym_pos`, `cholesky`, and possibly a list like `custom_solvers`), the list would be defined differently.  For instance, if `cholesky=False`, `cholmod` would not appear in the list; if `sparse=False`, the list would be made up of dense solvers. Any `custom_solvers` would be prepended to the list. 

`_get_delta` would no longer take in arguments `lstsq`, `sym_pos`, `cholesky`, or `permc_spec`. Instead, it would just take in the list `solvers`. It would start with the first element of the list and continue down the list in case of failure.

Does that make sense?

**Comment by hakeemo** [comment]
@mdhaber Yes makes sense.That is more powerful and (once implemented) easier to extend in my opinion.

**Comment by mdhaber** [comment]
Hey @hakeemo are you still interested in this? Thanks for your speedup of the sparse solver!

**Comment by mdhaber** [comment]
Hey @hakeemo just thought I'd check in again. Need anything from me?

**Comment by hakeemo** [comment]
@mdhaber sorry about being unresponsive. Yes , still interested in getting this done but don't see myself working on this for the next 4 weeks.

**Comment by mdhaber** [comment]
No problem; it's a crazy time. I'll check in next month!
